File: java/Debugger.java
package zxf;

import java.util.Arrays;
import java.net.InetAddress;
import java.util.concurrent.TimeUnit;
import java.awt.*;
import javax.script.Bindings;
import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.script.ScriptException;

public class Debugger {
    public static void main(String[] args) throws Exception {
      //sleep 1s
      Thread.sleep(1000);

      Class.forName("zzz.StaticLoader",false, Debugger.class.getClassLoader());

      //ouput log
      System.out.print("hello");
      System.out.println(" davis");

      //read env virable
      String path = System.getenv("PATH");
      System.out.println("path: " + path);

      //interrupt

      //run shell
      try {
          Process process = Runtime.getRuntime().exec("env");
          if (process.waitFor(30, TimeUnit.MINUTES)) {
              System.out.println("exit code: " + process.exitValue());
          } else {
              System.out.println("Timeout");
          }
      } catch (Exception ex) {
          ex.printStackTrace();
      }

      //out, err = exec.Command("ls","-al",".").Output()
      //if err != nil {
      //   log.Fatal(err)
      //}
      //fmt.Printf("The files is %s\n", out)

      //get stack info
      StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
      Arrays.stream(stackTrace).forEach(System.out::println);

      //profile

      //read config from external

      //format by template

      //pid In Java 9
      //System.out.print(ProcessHandle.current().pid());

      //thread
      System.out.println("Thread: " + Thread.currentThread());

      //hostname
      System.out.println("HostName: " + InetAddress.getLocalHost().getHostName());

      //IPs
      Enumeration e = NetworkInterface.getNetworkInterfaces();
      while(e.hasMoreElements())
      {
          NetworkInterface n = (NetworkInterface) e.nextElement();
          Enumeration ee = n.getInetAddresses();
          while (ee.hasMoreElements())
          {
              InetAddress i = (InetAddress) ee.nextElement();
              System.out.println(i.getHostAddress());
          }
      }

      //cpu info
      int cores = Runtime.getRuntime().availableProcessors();
      System.out.println("cores: " + cores);

      //cpu usage


      //dynamic load

      //proxy

      //host script
      ScriptEngine engine = new ScriptEngineManager().getEngineByExtension("js");
      Bindings bindings = engine.createBindings();
      bindings.put("a", 3600);
      bindings.put("b", 14);
      bindings.put("c", 4);
      try {
          Object d = engine.eval("a * b * c", bindings);
          System.out.println(d);
      } catch (ScriptException e) {
          e.printStackTrace();
      }

      //cwd
      System.out.println("user.dir: " + System.getProperty("user.dir"));

      //font && font-family
      GraphicsEnvironment e = GraphicsEnvironment.getLocalGraphicsEnvironment();
      Font[] fonts = e.getAllFonts();
      System.out.println("(name, family, isBold, isItalic, isPlain)");
      for (int i = 0; i < fonts.length; i++) {
          System.out.println(String.format("(%s, %s, %s, %s, %s)", fonts[i].getFontName(), fonts[i].getFamily(), fonts[i].isBold(), fonts[i].isItalic(), fonts[i].isPlain()));
      }

      String[] fontFamilyNames = e.getAvailableFontFamilyNames();
      for (int i = 0; i < fontFamilyNames.length; i++) {
          System.out.println(fontFamilyNames[i]);
      }
    }

    public static String getIP() throws SocketException {
        for(Enumeration<NetworkInterface> en = NetworkInterface.getNetworkInterfaces(); en.hasMoreElements();) {
          NetworkInterface inetf = en.nextElement();
          for(Enumeration<InetAddress> enumIpAddr = inetf.getInetAddresses(); enumIpAddr.hasMoreElements();) {
             InetAddress inetAddress = enumIpAddr.nextElement();
             if (!inetAddress.isLoopbackAddress() && (inetAddress instanceof Inet4Address)) {
                return inetAddress.getHostAddress();
             }
          }
        }
        return null;
    }
}

File: python/debugger.py
# -*- coding:utf-8 -*-  

#sleep 10s
import time;
time.sleep(10);

#ouput log
print("test");

#read env virable

#interrupt
import pdb;
#程序将主动暂停并进入pdb调试模式，不需要python -m pdb。按c继续运行。
pdb.set_trace();

#run shell
import subprocess;
subprocess.call("env");
#error
#subprocess.call("ls","-al",".");

#get stack info
import traceback;
traceback.print_stack();

#profile
def func_for_profile():
    print("sleepping...........");
    time.sleep(20);
func_for_profile();

#read config from external

#format by template

#current pid
import os;
print("pid: " + str(os.getpid()));

#current process info

#other process info
p = psutil.Process(16031)
p.name()      #进程名
p.exe()         #进程的bin路径
p.cwd()        #进程的工作目录绝对路径
p.status()     #进程状态
p.create_time()  #进程创建时间
p.uids()      #进程uid信息
p.gids()      #进程的gid信息
p.cpu_times()    #进程的cpu时间信息,包括user,system两个cpu信息
p.cpu_affinity()  #get进程cpu亲和度,如果要设置cpu亲和度,将cpu号作为参考就好
p.memory_percent()  #进程内存利用率
p.memory_info()    #进程内存rss,vms信息
p.io_counters()    #进程的IO信息,包括读写IO数字及参数
p.connectios()    #返回进程列表
p.num_threads()  #进程开启的线程数

#thread
import threading;
print(threading.currentThread().name)
print(threading.currentThread().ident)

#cpu info


#cpu usage
import psutil;
print(psutil.cpu_percent())

#dynamic load

#proxy

#host script

#cwd
import os;
print("cwd: " + os.getcwd());

#uuid
import uuid
print(uuid.uuid1())

#byte <-> char encoding
'abc'.encode('utf-8')
b'\x31\x32\x33'.decode('utf-8')

#hex string(1 byte to 2 chars)

#base64 string(3 bytes to 4 chars)

